bug: dnspolicy record controller panics if no name on listener

create a gateway with a listener that has no hostname
DNSPolicy will not get reconciled as it hits a nil pointer

panic: runtime error: invalid memory address or nil pointer dereference [recovered]
kuadrant-operator-controller-manager-8566585ffc-gqr66 manager 	panic: runtime error: invalid memory address or nil pointer dereference
kuadrant-operator-controller-manager-8566585ffc-gqr66 manager [signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x15995dc]
kuadrant-operator-controller-manager-8566585ffc-gqr66 manager
kuadrant-operator-controller-manager-8566585ffc-gqr66 manager goroutine 661 [running]:
kuadrant-operator-controller-manager-8566585ffc-gqr66 manager sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
kuadrant-operator-controller-manager-8566585ffc-gqr66 manager 	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:111 +0x1a4
kuadrant-operator-controller-manager-8566585ffc-gqr66 manager panic({0x183a240?, 0x2fee490?})
kuadrant-operator-controller-manager-8566585ffc-gqr66 manager 	/usr/local/go/src/runtime/panic.go:770 +0x124
kuadrant-operator-controller-manager-8566585ffc-gqr66 manager github.com/kuadrant/kuadrant-operator/controllers.(*DNSPolicyReconciler).reconcileGatewayDNSRecords(0x400065ba10, {0x1e9bad8, 0x4000b04600}, 0x40004af380, 0x40004ae000)
kuadrant-operator-controller-manager-8566585ffc-gqr66 manager 	/workspace/controllers/dnspolicy_dnsrecords.go:60 +0x23c
kuadrant-operator-controller-manager-8566585ffc-gqr66 manager github.com/kuadrant/kuadrant-operator/controllers.(*DNSPolicyReconciler).reconcileDNSRecords(0x400065ba10, {0x1e9bad8, 0x4000b04600}, 0x40004ae000, 0x40007dc4b0)
kuadrant-operator-controller-manager-8566585ffc-gqr66 manager 	/workspace/controllers/dnspolicy_dnsrecords.go:38 +0x424
kuadrant-operator-controller-manager-8566585ffc-gqr66 manager github.com/kuadrant/kuadrant-operator/controllers.(*DNSPolicyReco

--- a/controllers/dnspolicy_dnsrecords.go
+++ b/controllers/dnspolicy_dnsrecords.go
@@ -55,13 +55,12 @@ func (r *DNSPolicyReconciler) reconcileGatewayDNSRecords(ctx context.Context, ga
 	}
 
 	log.V(3).Info("checking gateway for attached routes ", "gateway", gateway.Name)
-
 	for _, listener := range gateway.Spec.Listeners {
-		listenerHost := *listener.Hostname
-		if listenerHost == "" {
-			log.Info("skipping listener no hostname assigned", listener.Name, "in ns ", gateway.Namespace)
+		if listener.Hostname == nil || *listener.Hostname == "" {
+			log.Info("skipping listener no hostname assigned", "listener", listener.Name, "in ns ", gateway.Namespace)
 			continue
 		}
+
 		hasAttachedRoute := false
 		for _, statusListener := range gateway.Status.Listeners {
 			if string(listener.Name) == string(statusListener.Name) {

--- a/tests/common/dnspolicy/dnspolicy_controller_test.go
+++ b/tests/common/dnspolicy/dnspolicy_controller_test.go
@@ -80,6 +80,7 @@ var _ = Describe("DNSPolicy controller", func() {
 
 	It("should validate loadBalancing field correctly", func(ctx SpecContext) {
 		gateway = tests.NewGatewayBuilder("test-gateway", gatewayClass.Name, testNamespace).
+			WithHTTPListener(tests.ListenerNameOne, "").
 			WithHTTPListener(tests.ListenerNameOne, tests.HostTwo(domain)).Gateway
 
 		// simple should succeed
@@ -140,6 +141,7 @@ var _ = Describe("DNSPolicy controller", func() {
 	It("should validate provider ref field correctly", func(ctx SpecContext) {
 
 		gateway = tests.NewGatewayBuilder("test-gateway", gatewayClass.Name, testNamespace).
+			WithHTTPListener(tests.ListenerNameOne, "").
 			WithHTTPListener(tests.ListenerNameOne, tests.HostTwo(domain)).Gateway
 
 		// should not allow an empty providerRef list

--- a/tests/common/tlspolicy/tlspolicy_controller_test.go
+++ b/tests/common/tlspolicy/tlspolicy_controller_test.go
@@ -115,6 +115,7 @@ var _ = Describe("TLSPolicy controller", func() {
 
 			By("creating a valid Gateway")
 			gateway = tests.NewGatewayBuilder("test-gateway", gatewayClass.Name, testNamespace).
+				WithHTTPListener("no-host", "").
 				WithHTTPListener("test-listener", "test.example.com").Gateway
 			Expect(k8sClient.Create(ctx, gateway)).To(Succeed())
 

--- a/tests/commons.go
+++ b/tests/commons.go
@@ -548,14 +548,20 @@ func (t *GatewayBuilder) WithLabels(labels map[string]string) *GatewayBuilder {
 	return t
 }
 
-func (t *GatewayBuilder) WithHTTPListener(name, hostname string) *GatewayBuilder {
-	typedHostname := gatewayapiv1.Hostname(hostname)
-	t.WithListener(gatewayapiv1.Listener{
+func (t *GatewayBuilder) WithHTTPListener(name string, hostname string) *GatewayBuilder {
+
+	var typedHostname gatewayapiv1.Hostname
+	gl := gatewayapiv1.Listener{
 		Name:     gatewayapiv1.SectionName(name),
-		Hostname: &typedHostname,
 		Port:     gatewayapiv1.PortNumber(80),
 		Protocol: gatewayapiv1.HTTPProtocolType,
-	})
+	}
+	if hostname != "" {
+		typedHostname = gatewayapiv1.Hostname(hostname)
+		gl.Hostname = &typedHostname
+	}
+
+	t.WithListener(gl)
 	return t
 }
 

